lib/Arch/Arch.cpp

```diff
@@ -0,0 +1,185 @@
+          }
+        } else if (auto value_intrin =
+                       llvm::dyn_cast<llvm::DbgValueInst>(debug_inst)) {
+          const llvm::MDNode *mdn = value_intrin->getVariable();
+          const llvm::DILocalVariable *div =
+              llvm::cast<llvm::DILocalVariable>(mdn);
+
+          if (div->getArg() != 0) {
+            param_names[div->getArg()] = div->getName().data();
+          }
+        }
+      }
+    }
+  }
+
+  // If we don't have names for some parameters then automatically name them
+  unsigned int num_args =
+      (unsigned int)(function.args().end() - function.args().begin());
+  for (unsigned int i = 1; i <= num_args; i++) {
+    if (!param_names.count(i)) {
+      param_names[i] = "param" + std::to_string(i);
+    }
+  }
+
+  return param_names;
+}
+
+// Try to allocate a register for the argument based on the register constraints
+// and what has already been reserved. Return nullptr if there is no possible
+// register allocation.
+const remill::Register *CallingConvention::TryRegisterAllocate(
+    llvm::Type &type, std::vector<bool> &reserved,
+    const std::vector<RegisterConstraint> &register_constraints) {
+  SizeConstraint size_constraint;
+  TypeConstraint type_constraint;
+
+  switch (type.getTypeID()) {
+    case llvm::Type::IntegerTyID: {
+      type_constraint = kTypeInt;
+      auto derived = llvm::cast<llvm::IntegerType>(type);
+      unsigned int width = derived.getBitWidth();
+      if (width <= 8) {
+        size_constraint = kMinBit8;
+      } else if (width <= 16) {
+        size_constraint = kMinBit16;
+      } else if (width <= 32) {
+        size_constraint = kMinBit32;
+      } else if (width <= 64) {
+        size_constraint = kMinBit64;
+      } else if (width <= 80) {
+        size_constraint = kMinBit80;
+      } else if (width <= 128) {
+        size_constraint = kMinBit128;
+      } else {
+        LOG(FATAL) << "Integer too big: "
+                   << remill::LLVMThingToString(&derived);
+      }
+      break;
+    }
+    case llvm::Type::FloatTyID: {
+      type_constraint = kTypeFloat;
+      // We automatically know it is 32-bit IEEE floating point type
+      size_constraint = kMinBit32;
+      break;
+    }
+    case llvm::Type::DoubleTyID: {
+      type_constraint = kTypeFloat;
+      // We automatically know it is 64-bit IEEE floating point type
+      size_constraint = kMinBit64;
+      break;
+    }
+    case llvm::Type::PointerTyID: {
+      type_constraint = kTypeIntegral;
+      size_constraint = kMinBit64;
+      break;
+    }
+    case llvm::Type::X86_FP80TyID: {
+      type_constraint = kTypeIntegral;
+      size_constraint = kMinBit80;
+      break;
+    }
+    case llvm::Type::VectorTyID: {
+      type_constraint = kTypeFloatOrVec;
+      size_constraint = kMinBit80;
+      break;
+    }
+    default: {
+      LOG(FATAL) << "Could not assign type and size constraints in "
+                    "TryRegisterAllocate()";
```

Comment:
Add in a `remill::LLVMThingToString` into the log message so that it prints out what type wasn't handled.

Reply:
Fixed e40ab35